common/src/main/java/io/pravega/common/concurrent/ScheduledQueue.java

```diff
@@ -0,0 +1,350 @@
+/**
+ * Copyright Pravega Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.pravega.common.concurrent;
+
+import com.google.common.collect.Iterators;
+import java.util.AbstractQueue;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.Nonnull;
+import lombok.Data;
+import lombok.val;
+
+/**
+ * Provides a unbounded blocking queue which {@link Scheduled} items can be added to.
+ * Items which are scheduled will not be returned from {@link #poll()} or {@link #take()} until their scheduled time.
+ * 
+ * This class is similar to DelayQueue but it allows a delay to be optional. This allows adding and polling non-delayed tasks in O(1).
+ * It also it lock-free, and may offer higher throughput under contention. 
+ */
+public class ScheduledQueue<E extends Scheduled> extends AbstractQueue<E> implements BlockingQueue<E> {
+   
+    private final AtomicLong itemsAdded = new AtomicLong(0);
```

Comment:
why are we counting all the items ever added and removed. 
i know that Long.Max is a very large value and we will not exhaust this practically. 
but can we do better?